Cache struct/class field offsets in SIL.

atrick opened the pull request:
The field's ordinal value is used by the Projection abstraction, which is
the basis of efficiently comparing and sorting access paths in SIL. It must
be cached before it is used by any SIL passes, including the verifier, or it
causes widespread quadratic complexity.

Fixes <rdar://problem/50353228> Swift compile time regression with optimizations enabled

In production code, a file that was taking 40 minutes to compile now
takes 1 minute, with more than half of the time in LLVM.

Here's a short script that reproduces the problem. It used to take 30s and now takes 0.06s:

```
// ./genlazyinit.swift > lazyinit.sil
// sil-opt ./lazyinit.sil --access-enforcement-opts

var NumProperties = 300

print("""
      sil_stage canonical

      import Builtin
      import Swift
      import SwiftShims

      public class LazyProperties {
      """)

for i in 0..<NumProperties {
  print("""
          //  public lazy var i\(i): Int { get set }
          @_hasStorage @_hasInitialValue final var __lazy_storage__i\(i): Int? { get set }
        """)
}

print("""
      }

     // LazyProperties.init()
     sil @$s4lazy14LazyPropertiesCACycfc : $@convention(method) (@owned LazyProperties) -> @owned LazyProperties {
     bb0(%0 : $LazyProperties):
       %enum = enum $Optional<Int>, #Optional.none!enumelt
     """)

for i in 0..<NumProperties {
  let adr = (i*4) + 2
  let access = adr + 1
  print("""
          %\(adr) = ref_element_addr %0 : $LazyProperties, #LazyProperties.__lazy_storage__i\(i)
          %\(access) = begin_access [modify] [dynamic] %\(adr) : $*Optional<Int>
          store %enum to %\(access) : $*Optional<Int>
          end_access %\(access) : $*Optional<Int>
        """)
}

print("""
        return %0 : $LazyProperties
      } // end sil function '$s4lazy14LazyPropertiesCACycfc'
      """)
```

atrick committed:
Cache struct/class field offsets in SIL.

The field's ordinal value is used by the Projection abstraction, which is
the basis of efficiently comparing and sorting access paths in SIL. It must
be cached before it is used by any SIL passes, including the verifier, or it
causes widespread quadratic complexity.

Fixes <rdar://problem/50353228> Swift compile time regression with optimizations enabled

In production code, a file that was taking 40 minutes to compile now
takes 1 minute, with more than half of the time in LLVM.

Here's a short script that reproduces the problem. It used to take 30s
and now takes 0.06s:

// swift genlazyinit.swift > lazyinit.sil
// sil-opt ./lazyinit.sil --access-enforcement-opts

var NumProperties = 300

print("""
      sil_stage canonical

      import Builtin
      import Swift
      import SwiftShims

      public class LazyProperties {
      """)

for i in 0..<NumProperties {
  print("""
          //  public lazy var i\(i): Int { get set }
          @_hasStorage @_hasInitialValue final var __lazy_storage__i\(i): Int? { get set }
        """)
}

print("""
      }

     // LazyProperties.init()
     sil @$s4lazy14LazyPropertiesCACycfc : $@convention(method) (@owned LazyProperties) -> @owned LazyProperties {
     bb0(%0 : $LazyProperties):
       %enum = enum $Optional<Int>, #Optional.none!enumelt
     """)

for i in 0..<NumProperties {
  let adr = (i*4) + 2
  let access = adr + 1
  print("""
          %\(adr) = ref_element_addr %0 : $LazyProperties, #LazyProperties.__lazy_storage__i\(i)
          %\(access) = begin_access [modify] [dynamic] %\(adr) : $*Optional<Int>
          store %enum to %\(access) : $*Optional<Int>
          end_access %\(access) : $*Optional<Int>
        """)
}

print("""
        return %0 : $LazyProperties
      } // end sil function '$s4lazy14LazyPropertiesCACycfc'
      """)

atrick commented:
@swift-ci test.

atrick commented:
@swift-ci test source compatibility.

atrick commented:
@swift-ci test compiler performance

atrick commented:
@swift-ci smoke test

gottesmm commented:
@atrick Can you put in a scale unit test for this with the test program? I think I remember how to do it if you do not.

gottesmm commented on include/swift/SIL/SILInstruction.h:
```diff
@@ -5625,39 +5625,62 @@ class TupleElementAddrInst
+/// be cached before it is used by any SIL passes, including the verifier, or it
+/// will cause widespread quadratic complexity. The type system does not (yet)
+/// cache a VarDecl's field index because it is unstable until the type is
+/// finalized.
```
This comment has a lot of implicit knowledge about SIL/SILOptimizer how it is used in it. I think we should be more explicit. My thoughts:

1. I would make it clear that the intent is for this to be used with "projection-like" instructions in the \brief.
2. I would make the 2nd sentence of the descriptive have more context especially around the type system comment.

My thought:

```
A common base for Projection-compatible instructions that use a cached field index to avoid quadratic complexity.

The Projection abstraction uses a compatible instruction's ordinal value to efficiently compare and sort access paths in SIL. Without using a cache, we must iterate over parts of the AST to find such an index resulting in widespread quadratic complexity.

FIXME: The reason why this is done is that the Decl TypeChecker does not currently cache a VarDecl's field index due to a VarDecl's field index being unstable until the type is finalized. If/when this is changed, we can eliminate this class and just access the pre-computed index on the actual VarDecl.

atrick replied:
Thanks. The wording I chose below is deliberate. If it doesn't make sense to you let me know.
```
/// A common base for instructions that require a cached field index.
///
/// "Field" is a term used here to refer to the ordered, accessible stored
/// properties of a class or struct.
///
/// The field's ordinal value is the basis of efficiently comparing and sorting
/// access paths in SIL. For example, whenever a Projection object is created,
/// is stores the field index. Finding the field index initially requires
/// searching the type declaration's array of all stored properties. If this
/// index is not cached, it will cause widespread quadratic complexity in any
/// pass that queries projections, including the SIL verifier.
///
/// FIXME: This cache may not be necessary if the Decl TypeChecker instead
/// caches a field index in the VarDecl itself. This solution would be superior
/// because it would allow constant time lookup of either the VarDecl or the
/// index from a single pointer without referring back to a projection
/// instruction.
```

gottesmm replied:
Amazing! +1! Only one typo:

```
For example, whenever a Projection object is created, is stores the field index.
```

=>

```
For example, whenever a Projection object is created, it stores the field index.
```

gottesmm commented on include/swift/SIL/Projection.h:
```diff
@@ -289,6 +289,8 @@ class Projection {
   /// Apply this projection to \p BaseType and return the relevant subfield's
   /// SILType if BaseField has less subtypes than projection's offset.
+  ///
+  /// WARNING: This is not a constant time operation.
```
Can you add here that it is not a constant time operation until we fix the typechecker? I am scared that this comment will be missed when the type checker is fixed. With that comment, it will be obvious that the comment is out of date.

atrick replied:
No, the implementation of this method itself isn't constant-time, regardless of the type checker's implementation. Code that needs to scale should find a way not to call it.

gottesmm replied:
From talking with Andy offline, the reason he said that this is not constant time is b/c of getVarDecl requesting all of the stored properties. I asked him to include that explicitly in this warning.

gottesmm commented on include/swift/SIL/Projection.h:
```diff
@@ -314,6 +316,7 @@ class Projection {
     llvm_unreachable("Unhandled ProjectionKind in switch.");
   }
 
+  /// WARNING: This is not a constant time operation.
```
Same here as above.

gottesmm reviewed (commented):
Overall looks ok. Some small asks.

compnerd commented on include/swift/SIL/SILInstruction.h:
```diff
@@ -5625,39 +5625,62 @@ class TupleElementAddrInst
+/// cache a VarDecl's field index because it is unstable until the type is
+/// finalized.
+class FieldIndexCacheBase : public SingleValueInstruction {
+  enum { InvalidFieldIndex = ~unsigned(0) };
```
This should be fine with MSVC.  The only thing which could be a problem is if you need it to be a fixed size (this will be a 4-byte enumeration).

atrick replied:
Yes, I should have used a fixed size enum.

gottesmm commented on include/swift/SIL/SILInstruction.h:
```diff
@@ -5670,71 +5702,33 @@ class StructExtractInst
-    auto s = getOperand()->getType().getStructOrBoundGenericStruct();
-    assert(s);
-    return s;
+    return cast<StructDecl>(getParentDecl());
```
One last question. Did you run this through git-clang-format? The reason I am asking is I am pretty sure it would put this on the same line as getStructDecl() [could be wrong though].

atrick replied:
I'm not sure. I was only using git format selectively because it keeps messing up the bitfield macros. When I ran it again though it didn't catch this for some reason.

gottesmm replied:
Ok. No worries! SGTM!

gottesmm commented on include/swift/SIL/SILInstruction.h:
```diff
@@ -5625,39 +5625,71 @@ class TupleElementAddrInst
-  StructDecl *getStructDecl() const {
-    auto s = getOperand()->getType().getStructOrBoundGenericStruct();
+  NominalTypeDecl *getParentDecl() const {
+    auto s = getOperand(0)->getType().getNominalOrBoundGenericNominal();
```
Meta question. Should the base type of this be some sort of unary instruction base or something? Then getOperand() is correct.

atrick replied:
Thanks for pointing that out; there should be an assert. There was actually a reason I didn't make this a subclass ofUnary Instruction.
https://github.com/apple/swift/pull/24717/commits/d98cf0cc09ca851632b3d064c985dca08f330b3e

atrick commented:
@gottesmm you can see review feedback in separate commits. I think that's easier to keep track of. I'm still working on the scale test. It's tricky because there's no sil-opt support yet. So I can count the number of calls to getStoredProperties, but that's probably still quadratic *somewhere* in the compiler. I can count the number of times SIL calls it from the new API, but that's useless as a regression test.

atrick committed:
Add scale-test --save-temps option.

I'm not sure how anyone debugs these tests otherwise.

atrick committed:
Add a scale-test for VarDecl::getStoredProperties.

Adds a NumStoredPropertiesQueries stat.

Adds a test case for an increasing number of lazy stored class
properties. Each property requires a formal access within the
initializer. This would manifest as cubic behavior in
AccessEnforcementOpts, which scales as O(1.5) in the above stat.

atrick commented:
@gottesmm I squashed the review feedback and added the scale test so it's ready to merge.

atrick commented:
@swift-ci test.
